File: backend/src/services/chat_history_service.py
from typing import List, Optional
from datetime import datetime
import uuid
import logging

from src.repositories.chat_thread_repository import ChatThreadRepository
from src.repositories.checkpoint_repository import CheckpointWriteRepository, CheckpointRepository
from src.repositories.message_content_repository import MessageContentRepository
from src.models.chat_models import (
    ChatThread, 
    ChatMessage, 
    ChatThreadSummary,
    ChatThreadWithMessages,
    CreateChatRequest
)
from src.services.checkpoint_service import CheckpointService
from src.repositories.messages_repository import MessagesRepository
from src.services.message_management_service import MessageManagementService

logger = logging.getLogger(__name__)

class ChatHistoryService:
    """Service for managing chat history using repository pattern"""
    
    def __init__(self, 
                 chat_thread_repo: ChatThreadRepository,
                 checkpoint_service: CheckpointService,
                 messages_repo: MessagesRepository,
                 message_content_repo: Optional[MessageContentRepository] = None):
        self.chat_thread_repo = chat_thread_repo
        self.checkpoint_service = checkpoint_service
        self.messages_repo = messages_repo
        self.message_content_repo = message_content_repo
    
    
    async def create_thread(self, request: CreateChatRequest, user_id: Optional[str] = None) -> ChatThread:
    
        try:
            # Generate thread_id
            thread_id = str(uuid.uuid4())
            
            # Create thread object
            thread = ChatThread(
                thread_id=thread_id, 
                title=request.title or "New Chat",
                created_at=datetime.now(),
                updated_at=datetime.now(),
                user_id=user_id  # Include user_id
            )
            
            if user_id:
                logger.info(f"Creating new chat thread: {thread_id} with user_id: {user_id}")
            
            # Save thread first
            success = await self.chat_thread_repo.create_thread(thread)
            if not success:
                raise Exception("Failed to create chat thread in database")
        
            
            logger.info(f"Created new chat thread: {thread_id}")
            
            # Verify thread was actually created
            try:
                created_thread = await self.chat_thread_repo.find_by_id(thread_id, "thread_id")
                logger.info(f"Thread verification - exists: {created_thread is not None}")
            except Exception as e:
                logger.error(f"Error verifying thread creation: {e}")
            
            return thread
            
        except Exception as e:
            logger.error(f"Error creating chat thread: {e}")
            raise Exception(f"Failed to create chat thread: {e}")
    
    async def get_thread(self, thread_id: str, user_id: Optional[str] = None) -> Optional[ChatThreadWithMessages]:
        try:
            thread = await self.chat_thread_repo.find_by_thread_id(thread_id, user_id=user_id)
            if not thread:
                return None
            
            messages = await self.get_thread_messages(thread_id)
            return ChatThreadWithMessages(
                thread_id=thread.thread_id,
                title=thread.title,
                created_at=thread.created_at,
                updated_at=thread.updated_at,
                user_id=thread.user_id,
                messages=messages
            )
        except Exception as e:
            logger.error(f"Error retrieving chat thread {thread_id}: {e}")
            raise Exception(f"Failed to retrieve chat thread: {e}")

    async def get_thread_messages(self, thread_id: str) -> List[ChatMessage]:
        try:
            messages = await self.messages_repo.get_all_messages_by_thread(thread_id)
            
            # Load content blocks for each message if message_content_repo is available
            if self.message_content_repo:
                for message in messages:
                    if message.message_id:
                        try:
                            content_blocks = await self.message_content_repo.get_blocks_by_message_id(message.message_id)
                            message.content = content_blocks
                        except Exception as e:
                            logger.warning(f"Failed to load content blocks for message {message.message_id}: {e}")
                            message.content = []
            
            return messages
        except Exception as e:
            logger.error(f"Error retrieving chat thread messages {thread_id}: {e}")
            raise Exception(f"Failed to retrieve chat thread messages: {e}")

    
    

    # Legacy update_message_flags method removed - use block-level approval instead

    async def get_message_by_id(self, thread_id: str, message_id: int) -> Optional[ChatMessage]:
        """Get a specific message by its ID within a thread."""
        try:
            return await self.messages_repo.get_message_by_id(thread_id, message_id)
        except Exception as e:
            logger.error(f"Error retrieving message {message_id} from thread {thread_id}: {e}")
            raise Exception(f"Failed to retrieve message: {e}")
    
    async def get_all_threads(self, limit: int = 50, skip: int = 0, user_id: Optional[str] = None) -> List[ChatThread]:
        try:
            return await self.chat_thread_repo.get_threads(limit=limit, skip=skip, user_id=user_id)
        except Exception as e:
            logger.error(f"Error retrieving chat threads: {e}")
            raise Exception(f"Failed to retrieve chat threads: {e}")

    async def get_all_threads_summary(self, limit: int = 50, skip: int = 0, user_id: Optional[str] = None) -> List[ChatThreadSummary]:
        try:
            chat_threads = await self.chat_thread_repo.get_threads(limit=limit, skip=skip, user_id=user_id)

            thread_summaries = []
            for thread in chat_threads:
                message_count = await self.messages_repo.count_messages_by_thread(thread.thread_id)
                last_message_obj = await self.messages_repo.get_last_message_by_thread(thread.thread_id)
                
                # Extract text preview from content blocks
                last_message = None
                if last_message_obj:
                    # Load content blocks if message_content_repo is available
                    if self.message_content_repo and last_message_obj.message_id:
                        try:
                            content_blocks = await self.message_content_repo.get_blocks_by_message_id(last_message_obj.message_id)
                            last_message_obj.content = content_blocks
                        except Exception as e:
                            logger.warning(f"Failed to load content blocks for message {last_message_obj.message_id}: {e}")
                            last_message_obj.content = []
                    
                    # Extract text from content blocks
                    if last_message_obj.content:
                        text_parts = []
                        for block in last_message_obj.content:
                            if isinstance(block, dict) and block.get('type') == 'text':
                                text = block.get('data', {}).get('text', '')
                                if text:
                                    text_parts.append(text)
                        
                        if text_parts:
                            # Join all text parts and truncate to 100 chars for preview
                            preview = ' '.join(text_parts)
                            last_message = preview[:100] + '...' if len(preview) > 100 else preview
                
                thread_summary = ChatThreadSummary(
                    thread_id=thread.thread_id,
                    title=thread.title,
                    created_at=thread.created_at,
                    updated_at=thread.updated_at,
                    last_message=last_message,
                    message_count=message_count
                )
                thread_summaries.append(thread_summary)
            return thread_summaries
        except Exception as e:
            logger.error(f"Error retrieving chat thread summaries: {e}")
            raise Exception(f"Failed to retrieve chat thread summaries: {e}")



    
    async def delete_thread(self, thread_id: str) -> bool:
    
        try:
            # First, get all message IDs for this thread to clean up message_content
            messages = await self.messages_repo.get_all_messages_by_thread(thread_id)
            message_ids = [msg.message_id for msg in messages if msg.message_id]
            
            # Delete all message_content blocks for these messages
            if message_ids and self.message_content_repo:
                try:
                    total_blocks_deleted = 0
                    for message_id in message_ids:
                        deleted_count = await self.message_content_repo.delete_blocks_by_message_id(message_id)
                        total_blocks_deleted += deleted_count
                    if total_blocks_deleted > 0:
                        logger.info(f"Deleted {total_blocks_deleted} content blocks for {len(message_ids)} messages in thread {thread_id}")
                except Exception as e:
                    logger.warning(f"Failed to delete message content blocks for thread {thread_id}: {e}")
                    # Don't fail the whole operation if content block cleanup fails
            
            # Delete the messages
            await self.messages_repo.delete_messages_by_thread(thread_id)
            
            # Delete the thread
            thread_deleted = await self.chat_thread_repo.delete_thread(thread_id)
            
            if thread_deleted:
                logger.info(f"Deleted chat thread: {thread_id}")
                
                # Clean up associated checkpoint data
                try:
                    checkpoint_result = await self.checkpoint_service.delete_all_thread_data(thread_id)
                    total_deleted = checkpoint_result.get('total_deleted', 0)
                    if total_deleted > 0:
                        logger.info(f"Deleted {total_deleted} checkpoint records for thread {thread_id}")
                except Exception as e:
                    logger.warning(f"Failed to delete checkpoint data for thread {thread_id}: {e}")
                    # Don't fail the whole operation if checkpoint cleanup fails
                
                return True
            else:
                logger.warning(f"Thread {thread_id} not found for deletion")
                return False
                
        except Exception as e:
            logger.error(f"Error deleting chat thread {thread_id}: {e}")
            raise Exception(f"Failed to delete chat thread: {e}")
    
    async def update_thread_title(self, thread_id: str, title: str) -> bool:
    
        try:
            success = await self.chat_thread_repo.update_thread_title(thread_id, title)
            if success:
                logger.info(f"Updated title for thread {thread_id}")
            else:
                logger.warning(f"Thread {thread_id} not found for title update")
            return success
        except Exception as e:
            logger.error(f"Error updating thread title {thread_id}: {e}")
            raise Exception(f"Failed to update thread title: {e}")
    
    async def get_thread_count(self, user_id: Optional[str] = None) -> int:
    
        try:
            return await self.chat_thread_repo.count_threads(user_id=user_id)
        except Exception as e:
            logger.error(f"Error counting chat threads: {e}")
            return 0

File: frontend/src/api/services/chatHistoryService.ts
import { apiClient } from '../client';
import { ChatMessage } from '../types';

export interface ChatThread {
  thread_id: string;
  title?: string;
  messages: ChatMessage[];
  created_at: string;
  updated_at: string;
}

export interface ChatThreadSummary {
  thread_id: string;
  title?: string;
  last_message?: string;
  message_count: number;
  created_at: string;
  updated_at: string;
}

export interface CreateChatRequest {
  title?: string;
  initial_message?: string;
}

export interface AddMessageRequest {
  thread_id: string;
  sender: 'user' | 'assistant';
  content: string;
  content_blocks?: Array<{
    id: string;
    type: 'text' | 'tool_calls' | 'explorer' | 'visualizations';
    needsApproval?: boolean;
    data: any;
  }>;
  message_type?: 'message' | 'explorer' | 'visualization' | 'structured';
  checkpoint_id?: string;
  
  // Additional fields from frontend Message interface
  message_id?: number;
  needs_approval?: boolean;
  approved?: boolean;
  disapproved?: boolean;
  is_error?: boolean;
  is_feedback?: boolean;
  has_timed_out?: boolean;
  can_retry?: boolean;
  retry_action?: 'approve' | 'feedback' | 'cancel';
  thread_id_ref?: string;
}

export interface ChatHistoryResponse {
  success: boolean;
  data?: ChatThread;
  message: string;
}

export interface ChatListResponse {
  success: boolean;
  data: ChatThreadSummary[];
  message: string;
  total: number;
}

export class ChatHistoryService {
  private static client = apiClient;

  /**
   * Create a new chat thread
   */
  static async createThread(request: CreateChatRequest): Promise<ChatThread> {
    const response = await this.client.post<ChatHistoryResponse>('/chat-history/create', request);
    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to create chat thread');
    }
    return response.data;
  }

  /**
   * Get all chat threads
   */
  static async getAllThreads(limit = 50, skip = 0): Promise<{ threads: ChatThreadSummary[], total: number }> {
    const response = await this.client.get<ChatListResponse>(`/chat-history/threads?limit=${limit}&skip=${skip}`);
    if (!response.success) {
      throw new Error(response.message || 'Failed to get chat threads');
    }
    return { threads: response.data, total: response.total };
  }

  /**
   * Get a specific chat thread with full history
   */
  static async getThread(threadId: string): Promise<ChatThread> {
    const response = await this.client.get<ChatHistoryResponse>(`/chat-history/thread/${threadId}`);
    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to get chat thread');
    }
    return response.data;
  }

  /**
   * Add a message to an existing thread
   */
  static async addMessage(request: AddMessageRequest): Promise<void> {
    const response = await this.client.post<{ success: boolean; message: string }>('/chat-history/add-message', request);
    if (!response.success) {
      throw new Error(response.message || 'Failed to add message');
    }
  }

  /**
   * Restore a chat thread for continuing conversation
   */
  static async restoreThread(threadId: string): Promise<ChatThread> {
    const response = await this.client.post<ChatHistoryResponse>(`/chat-history/thread/${threadId}/restore`, {});
    if (!response.success || !response.data) {
      throw new Error(response.message || 'Failed to restore chat thread');
    }
    return response.data;
  }

  /**
   * Update thread title
   */
  static async updateThreadTitle(threadId: string, title: string): Promise<void> {
    const response = await this.client.put<{ success: boolean; message: string }>(`/chat-history/thread/${threadId}/title?title=${encodeURIComponent(title)}`, {});
    if (!response.success) {
      throw new Error(response.message || 'Failed to update thread title');
    }
  }

  /**
   * Delete a chat thread
   */
  static async deleteThread(threadId: string): Promise<void> {
    const response = await this.client.delete<{ success: boolean; message: string }>(`/chat-history/thread/${threadId}`);
    if (!response.success) {
      throw new Error(response.message || 'Failed to delete chat thread');
    }
  }

  /**
   * Update message flags (e.g., approval) persistently
   * @deprecated This endpoint has been removed. Use updateBlockFlags for block-level updates instead.
   */
  static async updateMessageFlags(threadId: string, params: {
    message_id: number;
    needs_approval?: boolean;
    approved?: boolean;
    disapproved?: boolean;
    is_error?: boolean;
    is_feedback?: boolean;
    has_timed_out?: boolean;
    can_retry?: boolean;
    retry_action?: 'approve' | 'feedback' | 'cancel';
  }): Promise<{ success: boolean; message: string; updated_message?: any }> {
    // Legacy endpoint removed - this is a no-op for backward compatibility
    console.warn('updateMessageFlags is deprecated. Use updateBlockFlags for block-level updates instead.');
    return {
      success: true,
      message: 'Legacy endpoint removed - use block-level updates instead'
    };
  }

  /**
   * Update block-level status persistently
   */
  static async updateBlockFlags(
    threadId: string,
    messageId: number,
    blockId: string,
    params: {
      needsApproval?: boolean;
      messageStatus?: 'pending' | 'approved' | 'rejected' | 'error' | 'timeout';
    }
  ): Promise<{ success: boolean; message: string; updated_fields?: string[] }> {
    const response = await this.client.put<{ success: boolean; message: string; updated_fields?: string[] }>(
      `/chat-history/thread/${threadId}/message/${messageId}/block/${blockId}/approval`,
      params
    );
    if (!response.success) {
      throw new Error(response.message || 'Failed to update block flags');
    }
    return response;
  }
}
